File: packages/schemas/src/lesson.ts
import { z } from "zod";
import { SubjectEnum, DifficultyEnum, AgeGroupEnum, TimestampSchema } from "./common";

// Lesson content type
export const LessonContentSchema = z.object({
  type: z.enum(["text", "audio", "video", "interactive"]),
  content: z.string(),
  duration: z.number().optional(), // in minutes
  metadata: z.record(z.any()).optional()
});

// Activity within a lesson
export const ActivitySchema = z.object({
  id: z.string(),
  type: z.enum(["reading", "listening", "speaking", "writing", "quiz", "game"]),
  title: z.string(),
  description: z.string(),
  content: LessonContentSchema,
  expectedDuration: z.number(), // in minutes
  points: z.number().min(0).default(10),
  requiredForCompletion: z.boolean().default(true)
});

// Main lesson schema
export const LessonSchema = z.object({
  id: z.string().uuid(),
  title: z.string().min(3),
  description: z.string(),
  subject: SubjectEnum,
  ageGroup: AgeGroupEnum,
  difficulty: DifficultyEnum,
  
  // Content
  activities: z.array(ActivitySchema).min(1),
  estimatedDuration: z.number(), // total minutes
  
  // Learning objectives
  objectives: z.array(z.string()).min(1),
  keywords: z.array(z.string()).default([]),
  
  // Prerequisites and progression
  prerequisites: z.array(z.string()).default([]), // lesson IDs
  unlocks: z.array(z.string()).default([]), // lesson IDs
  
  // Metadata
  isPublished: z.boolean().default(false),
  tags: z.array(z.string()).default([])
}).merge(TimestampSchema);

// Create lesson request
export const CreateLessonSchema = LessonSchema.omit({
  id: true,
  createdAt: true,
  updatedAt: true
});

export type LessonContent = z.infer<typeof LessonContentSchema>;
export type Activity = z.infer<typeof ActivitySchema>;
export type Lesson = z.infer<typeof LessonSchema>;
export type CreateLessonRequest = z.infer<typeof CreateLessonSchema>;

File: backend/app/models/lesson.py
import json
from datetime import datetime
from typing import TYPE_CHECKING
from uuid import UUID, uuid4

from sqlmodel import Field, Relationship, SQLModel

if TYPE_CHECKING:
    from .session import Session

class Activity(SQLModel, table=True):
    __tablename__ = "activities"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    lesson_id: UUID = Field(foreign_key="lessons.id", index=True)
    activity_id: str  # Internal activity identifier
    type: str  # "reading", "listening", "speaking", "writing", "quiz", "game"
    title: str
    description: str
    content: str  # JSON content
    expected_duration: int  # minutes
    points: int = Field(default=10)
    required_for_completion: bool = Field(default=True)
    order_index: int = Field(default=0)

    # Relationships
    lesson: "Lesson" = Relationship(back_populates="activities")


class Lesson(SQLModel, table=True):
    __tablename__ = "lessons"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    title: str
    description: str
    subject: str  # "arabic", "english", "islamic"
    age_group: str  # "4-6", "7-9", "10-12"
    difficulty: str  # "beginner", "intermediate", "advanced"

    # Content
    estimated_duration: int  # total minutes
    objectives: str  # JSON array
    keywords: str = Field(default="[]")  # JSON array

    # Prerequisites and progression
    prerequisites: str = Field(default="[]")  # JSON array of lesson IDs
    unlocks: str = Field(default="[]")  # JSON array of lesson IDs

    # Metadata
    is_published: bool = Field(default=False)
    tags: str = Field(default="[]")  # JSON array

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime | None = Field(default=None)

    # Relationships
    activities: list[Activity] = Relationship(back_populates="lesson")
    sessions: list["Session"] = Relationship(back_populates="lesson")

    @property
    def objectives_list(self) -> list[str]:
        return json.loads(self.objectives)

    @property
    def keywords_list(self) -> list[str]:
        return json.loads(self.keywords)

    @property
    def prerequisites_list(self) -> list[str]:
        return json.loads(self.prerequisites)

    @property
    def unlocks_list(self) -> list[str]:
        return json.loads(self.unlocks)

    @property
    def tags_list(self) -> list[str]:
        return json.loads(self.tags)

    class Config:
        validate_assignment = True
